TSPL.docc/LanguageGuide/Generics.md

```diff
@@ -1853,6 +1853,286 @@ protocol ComparableContainer: Container where Item: Comparable { }
+
+- A type pack provides the types for a value pack.
+  The corresponding types and values appear at the same positions
+  in their respective packs.
+
+- When you write code that works on collections, you use iteration.
+  Working with parameter packs is similar ---
+  except each element has a different type,
+  and instead of iteration you use repetition.
+
+How do you create a parameter pack?
+
+- In the generic type parameters,
+  write `each` in front of a generic argument
+  to indicate that this argument creates a type parameter pack.
+
+- In the function's parameters,
+  write `repeat` in front of the type for the parameter
+  that can accept a variable number of arguments,
+  to create an *expansion pattern*.
+  You can also write `repeat` in the function's return type.
+
+- The expansion of a repetition pattern produces a comma-separated list.
+  It can appear in generic argument lists,
+  in tuples types,
+  and in function argument lists.
+
+  ```swift
+  func f<each T>(_ t: repeat each T) -> repeat each T
+  ```
+
+- The expansion pattern is repeated for every element in the given type pack
+  by iterating over the types in the type pack
+  and replacing the type placeholder that comes after `each`.
+
+  For example, expanding `repeat Request<each Payload>`
+  where the `Payload` type pack contains `Bool`, `Int`, and `String`
+  produces `Request<Bool>, Request<Int>, Request<String>`.
+
+- When the type of a function argument is a type pack,
+  the values that are passed in for that argument become a value pack.
+
+- Naming convention:
+  Use singular names for parameter packs,
+  and plural names only in argument labels.
+
+Note:
+Parameter packs can contain zero or more arguments.
+If you need to require one or more,
+use a regular parameter before the pack parameters.
+
+What else can you repeat?
+How do you repeat more than one type?
+
+- In the simple case, where only one type repeats,
+  you write `repeat each T` or similar.
+
+- For collections or other generic types,
+  the pack expansion can happen inside,
+  like `repeat Array<each T>` expands to multiple array types.
+
+- A more complex type can include `each` multiple times,
+  like `repeat Dictionary<each Key, each Value>`.
+  All of the expansions have to be the same size ---
+  in this example,
+  the list of types that `Key` expands to must be the same length
+  as the list that `Value` expands to.
+
+How do you vary the return types, based on the parameter types?
+(TODO: Worked code example.
+Use WWDC 2023 session 10168 example at 16 minutes
+as a starting point.)
+
+How do you constrain the types in a parameter pack?
+
+- In the simple case,
+  write the `where` clause in the generic parameters list.
+  `func foo<each T: Runcible>`
+  requires all of the types passed in the `T` type pack to conform.
+
+- In the more complex case,
+  use `repeat each T ` in a trailing `where` clause.
+
+- You must restrict the types that appear in a type-parameter pack;
```

Comment:
This seems a little over-general? Because you do not always need to restrict the types. "You must restrict the types that appear in a type parameter pack if its expansion will be used as a restricted generic type parameter" perhaps?